File: Leetcode 3101-3200/3149/3149.py
# Leetcode 3149: Find the Minimum Cost Array Permutation
# https://leetcode.com/problems/find-the-minimum-cost-array-permutation/
# Solved on 1st of December, 2025
class Solution:
    def findPermutation(self, nums: list[int]) -> list[int]:
        """
        Finds the permutation of indices that minimizes the total cost.
        The cost is defined as the sum of absolute differences between adjacent elements in the permutation,
        plus the absolute difference between the last and first elements.
        :param nums: A list of integers representing the values at each index.
        :return: A list of integers representing the permutation of indices.
        """
        n = len(nums)
        memo = [[-1] * n for _ in range(1 << n)]
        pathChoice = [[-1] * n for _ in range(1 << n)]

        def calculateMinCost(mask, lastIndex):
            if mask == (1 << n) - 1:
                return abs(lastIndex - nums[0])

            if memo[mask][lastIndex] != -1:
                return memo[mask][lastIndex]

            minCost = float('inf')
            bestNode = -1

            for nextIndex in range(n):
                if not (mask & (1 << nextIndex)):
                    currentCost = abs(lastIndex - nums[nextIndex]) + calculateMinCost(mask | (1 << nextIndex),
                                                                                      nextIndex)
                    if currentCost < minCost:
                        minCost = currentCost
                        bestNode = nextIndex

            memo[mask][lastIndex] = minCost
            pathChoice[mask][lastIndex] = bestNode
            return minCost

        calculateMinCost(1, 0)

        resultPermutation = [0]
        currentMask = 1
        currentIndex = 0

        for _ in range(n - 1):
            nextNode = pathChoice[currentMask][currentIndex]
            resultPermutation.append(nextNode)
            currentMask |= (1 << nextNode)
            currentIndex = nextNode

        return resultPermutation

File: Leetcode 3101-3200/3149/3149.java
// Leetcode 3149: Find the Minimum Cost Array Permutation
// https://leetcode.com/problems/find-the-minimum-cost-array-permutation/
// Solved on 1st of December, 2025
class Solution {
    private int[] nums;
    private int[][] memo;
    private int n;

    /**
     * Finds the permutation of `nums` that minimizes the total cost.
     * The cost is defined as the sum of absolute differences between adjacent elements in the permutation,
     * plus the absolute difference between the last and first elements.
     * @param nums The input array of integers.
     * @return An array representing the permutation with the minimum cost.
     */
    public int[] findPermutation(int[] nums) {
        this.nums = nums;
        this.n = nums.length;
        this.memo = new int[1 << n][n];
        
        for (int i = 0; i < (1 << n); i++) {
            for (int j = 0; j < n; j++) {
                memo[i][j] = -1;
            }
        }

        solve(1, 0);

        int[] res = new int[n];
        int mask = 1;
        int prev = 0;
        res[0] = 0;

        for (int i = 1; i < n; i++) {
            for (int next = 0; next < n; next++) {
                if ((mask & (1 << next)) == 0) {
                    int score = Math.abs(prev - nums[next]) + solve(mask | (1 << next), next);
                    if (score == memo[mask][prev]) {
                        res[i] = next;
                        mask |= (1 << next);
                        prev = next;
                        break;
                    }
                }
            }
        }
        return res;
    }

    private int solve(int mask, int prev) {
        if (mask == (1 << n) - 1) {
            return Math.abs(prev - nums[0]);
        }
        if (memo[mask][prev] != -1) {
            return memo[mask][prev];
        }

        int minScore = Integer.MAX_VALUE;
        for (int next = 0; next < n; next++) {
            if ((mask & (1 << next)) == 0) {
                int score = Math.abs(prev - nums[next]) + solve(mask | (1 << next), next);
                if (score < minScore) {
                    minScore = score;
                }
            }
        }
        return memo[mask][prev] = minScore;
    }
}